src/cpp/session/modules/clang/CodeCompletion.cpp

```diff
@@ -154,6 +169,237 @@ core::json::Object toJson(const CodeCompleteResult& result)
    return resultJson;
 }
 
+void discoverTranslationUnitIncludePaths(const FilePath& filePath,
+                                         std::vector<std::string>* pIncludePaths)
+{
+   std::vector<std::string> args =
+         rCompilationDatabase().compileArgsForTranslationUnit(
+            filePath.absolutePathNative(), false);
+   
+   BOOST_FOREACH(const std::string& arg, args)
+   {
+      if (boost::algorithm::starts_with(arg, "-I"))
+      {
+         // skip RStudio's libclang builtin header paths
+         if (arg.find("libclang/builtin-headers") != std::string::npos)
+            continue;
+         
+         pIncludePaths->push_back(
+                  string_utils::strippedOfQuotes(arg.substr(2)));
+      }
+   }
+}
+
+void discoverSystemIncludePaths(std::vector<std::string>* pIncludePaths)
+{
+   core::system::ProcessOptions processOptions;
+   processOptions.redirectStdErrToStdOut = true;
+   
+   // resolve R CMD location for shell command
+   FilePath rBinDir;
+   Error error = module_context::rBinDir(&rBinDir);
+   if (error)
+   {
+      LOG_ERROR(error);
+      return;
+   }
+   
+   shell_utils::ShellCommand rCmd = module_context::rCmd(rBinDir);
+   
+   // get the CXX compiler by asking R
+   std::string compilerPath;
+   {
+      core::system::ProcessResult result;
+      rCmd << "config";
+      rCmd << "CXX";
+      Error error = core::system::runCommand(rCmd, processOptions, &result);
+      if (error)
+         LOG_ERROR(error);
+      else if (result.exitStatus != EXIT_SUCCESS)
+         LOG_ERROR_MESSAGE("Error querying CXX compiler: " + result.stdOut);
+      else
+         compilerPath = string_utils::trimWhitespace(result.stdOut);
+   }
+   if (compilerPath.empty())
+      return;
+   
+   // ask the compiler what the system include paths are (note that both
+   // gcc and clang accept the same command)
+   std::string includePathOutput;
+   {
+      core::system::ProcessResult result;
+      std::string cmd = compilerPath + " -E -x c++ - -v < " kDevNull;
+      
+      // TODO: this needs to execute with Rtools on the PATH
```

Comment:
@jjallaire do we have a helper that runs a shell command after placing Rtools on the PATH?